src/Symfony/Component/Security/Core/User/UserProviderInterface.php

```diff
@@ -12,36 +12,40 @@
 namespace Symfony\Component\Security\Core\User;
 
 use Symfony\Component\Security\Core\Exception\UnsupportedUserException;
-use Symfony\Component\Security\Core\Exception\UsernameNotFoundException;
+use Symfony\Component\Security\Core\Exception\UserNotFoundException;
 
 /**
  * Represents a class that loads UserInterface objects from some source for the authentication system.
  *
- * In a typical authentication configuration, a username (i.e. some unique
- * user identifier) credential enters the system (via form login, or any
- * method). The user provider that is configured with that authentication
- * method is asked to load the UserInterface object for the given username
- * (via loadUserByUsername) so that the rest of the process can continue.
+ * In a typical authentication configuration, a user identifier (e.g. a
+ * username or e-mailaddress) credential enters the system (via form login, or
+ * any method). The user provider that is configured with that authentication
+ * method is asked to load the UserInterface object for the given identifier (via
+ * loadUserByIdentifier) so that the rest of the process can continue.
  *
  * Internally, a user provider can load users from any source (databases,
  * configuration, web service). This is totally independent of how the authentication
  * information is submitted or what the UserInterface object looks like.
  *
  * @see UserInterface
  *
+ * @method UserInterface loadUserByIdentifier(string $identifier) loads the user for the given user identifier (e.g. username or email)
+ *
  * @author Fabien Potencier <fabien@symfony.com>
  */
 interface UserProviderInterface
 {
     /**
      * Loads the user for the given username.
      *
-     * This method must throw UsernameNotFoundException if the user is not
+     * This method must throw UserNotFoundException if the user is not
      * found.
      *
      * @return UserInterface
      *
-     * @throws UsernameNotFoundException if the user is not found
+     * @throws UserNotFoundException if the user is not found
+     *
+     * @deprecated since Symfony 5.3, to be removed in 6.0, use loadUserByIdentifier() instead.
```

Comment:
This is likely to trigger unfixable notices as DebugClassLoader will ask to stop implementing  the method, isn't it ? Same for other interfaces?

Reply:
Hmm, but it's the only way to make sure deprecations are triggered when calling the method, right?

Reply:
I think that we cannot really deprecate an interface method. Triggering deprecation notices when calling it in core is the only thing we can do (you did it already in e.g. `AuthenticationProviderManager`). 
See https://github.com/symfony/symfony/pull/30388 for a precedent, we have to remove the method at some point (maybe now? /cc @nicolas-grekas).

Reply:
I think we should remove the method from the interface. If I want to implement a new user provider, I shouldn't be forced to implement a deprecated method.

Reply:
So to make it clear: Both the new and deprecated method in the interface (and the other modified interfaces) should be moved to an `@method` PHPdoc.

This would mean we loose the ability to trigger deprecations when using this deprecated method in your own implementation of the interface.
And, maybe even more important, it breaks all tests in applications that mock `UserProviderInterface::class`.

Reply:
The old method should not be moved to an `@method` phpdoc but just removed, as making it virtual means it should be implemented.

> to be clear: I'm not against doing it this way, but I'm not sure if the pros (keeping a deprecated method) are heavier than the cons (breaking tests while upgrading a minor).

I understand and share your concerns, but going with `@deprecated` is a no-go as it'd mean triggering a deprecation notice that is impossible to fix until 6.0 for all (existing and new) implementations, while a broken mock can be fixed. 
We already accepted this tradeoff in https://github.com/symfony/symfony/issues/31814#issuecomment-498756343

Reply:
Alright, thanks for the reference. I'll update the PR tonight.